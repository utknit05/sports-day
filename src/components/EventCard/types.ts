import { IEvent } from "../../context/EventsList/types";

type TEventActionHandler = (event: IEvent) => void

export type TEventCardProps =
    { event: IEvent } &
    ({
        type: 'selected';
        onRemove: TEventActionHandler;
    } | {
        type: 'unselected';
        onSelect: TEventActionHandler;
    })

export enum EEventsCategory {
    Swimming = 'Swimming',
    Athletics = 'Athletics',
    Boxing = 'Boxing',
}

export const categoryToGradientMap: {
    [key in EEventsCategory]: string
} = {
    [EEventsCategory.Swimming]: 'linear-gradient(to left, rgb(27, 214, 255), rgb(59, 157, 255))',
    [EEventsCategory.Athletics]: 'linear-gradient(to right, rgb(236, 97, 23, 90), rgb(0, 108, 255, 90))',
    [EEventsCategory.Boxing]: 'linear-gradient(to right, rgb(236, 58, 23), rgb(255, 255, 255))',
}
