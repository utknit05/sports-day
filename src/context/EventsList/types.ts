import { IListResponse } from "../../utils/api";
import { fetchEvents, fetchEventsCompleted, resetEvents } from "./store";

export const GET_EVENTS_ENDPOINT = 'https://run.mocky.io/v3/2744c231-8991-4ae8-bc45-1f645437585a';

export interface IEvent {
  id: number;
  event_name: string;
  event_category: string;
  start_time: string;
  end_time: string;
  disabled?: boolean;
}

export type TEventsState = IListResponse<IEvent>
export type TEventsActions = ReturnType<typeof fetchEvents | typeof fetchEventsCompleted | typeof resetEvents>
export interface IEventsContext {
  selectedEvents: IEvent[];
  unselectedEvents: IEvent[];
  selectEvent: (event: IEvent) => void
  removeEvent: (event: IEvent) => void
}
