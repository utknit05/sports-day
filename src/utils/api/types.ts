export enum ERequestState {
    INITIAL,
    REQUESTING,
    SUCCESS,
    FAILED,
}

export interface IListResponse<T>  {
    list: T[];
    requestState: ERequestState;
    error?: {
        code: string;
        message: string;
    };
}
