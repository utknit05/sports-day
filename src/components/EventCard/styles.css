.eventCardContainer {
    padding: 12px;
    border: 1px solid #000;
    border-radius: 6px;
    box-shadow: 2px 2px #cbcaca;
    max-width: 250px;
    background-image: linear-gradient(to right, #553c9a, #b393d3);
    justify-content: center;
}

.eventDetailsContainer {
    align-items: center;
    display: flex;
}

.eventDescription {
    border-left: 1px solid #000;
    padding: 16px 0 16px 16px;
    margin-left: 16px;
}

.eventName {
    font-size: 1.2rem;
    font-weight: 600;
    margin: 4px 0;
}

.eventSubText {
    font-size: 0.9rem;
    margin: 1px 0;
}

.eventButtonWrapper {
    display: flex;
    justify-content: flex-end;
}

.eventActionButton {
    border-radius: 6px;
    color: #000;
    padding: 6px 10px;
    margin-top: 8px;
    cursor: pointer;
}

.selectButton {
    background-color: #beefc0;
    border: 1px solid #279303;
}

.removeButton {
    background-color: #efcebe;
    border: 1px solid #932003;
}

.disabledButton {
    background-color: #ebefeb;
    border: 1px solid #070707;
}
