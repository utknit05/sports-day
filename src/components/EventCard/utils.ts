import { EEventsCategory, TEventCardProps, categoryToGradientMap } from "./types";

export const categoryGradientStyle = (props: TEventCardProps): React.CSSProperties => {
    if (props.type === 'unselected' && props.event.disabled) {
        return { backgroundImage: 'linear-gradient(to right, #9a9998, #ffffff)' };
    }

    return { backgroundImage: categoryToGradientMap[props.event.event_category as EEventsCategory] };
}

export const getButtonStyleClass = (props: TEventCardProps) => {
    if (props.type === 'selected') {
        return 'removeButton';
    }

    if (props.event.disabled) {
        return 'disabledButton';
    }

    return 'selectButton';
}
