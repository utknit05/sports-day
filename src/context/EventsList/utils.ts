import { IEvent } from "./types";

export const getDisabledEventsInUnselectedList = (selectedEvents: IEvent[], unselectedEvents: IEvent[], conflictingEventsMap?: { [key: number]: number[] }) => {
    if (!conflictingEventsMap) {
        return unselectedEvents;
    }

    let disabledEventsIds: number[] = [];
    selectedEvents.forEach(({ id }) => {
        conflictingEventsMap[id] && disabledEventsIds.push(...conflictingEventsMap[id]);
    })

    return unselectedEvents.map((event) => {
        if (disabledEventsIds.includes(event.id)) {
            return {
                ...event,
                disabled: true,
            }
        }

        return {
            ...event,
            disabled: false,
        };
    })
}
