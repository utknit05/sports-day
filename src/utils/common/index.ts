import { IEvent } from "../../context/EventsList/types";

export const returnMinutes = (time: number) => {
    if (time > 0 && time < 10) {
        return `:0${time}`;
    }

    if (time >= 10) {
        return `:${time}`;
    }

    return ''
}

export const getDuration = (startTimestamp: string, endTimestamp: string) => {
    const startDate = new Date(startTimestamp);
    const endDate = new Date(endTimestamp);

    if (isNaN(startDate.getTime()) || isNaN(endDate.getTime())) {
        return '';
    }

    const [startHours, startMinutes] = [startDate.getHours(), startDate.getMinutes()];
    const [endHours, endMinutes] = [endDate.getHours(), endDate.getMinutes()];

    const startAMOrPM = startHours > 12 ? 'PM' : 'AM';
    const endAMOrPM = endHours > 12 ? 'PM' : 'AM';

    const start12HoursClock = startHours > 12 ? startHours % 12 : startHours;
    const end12HoursClock = endHours > 12 ? endHours % 12 : endHours;

    return `${start12HoursClock}${returnMinutes(startMinutes)} ${startAMOrPM} - ${end12HoursClock}${returnMinutes(endMinutes)} ${endAMOrPM}`;
}

export const compareDates = (timestamp1: string, timestamp2: string) => {
    const date1 = new Date(timestamp1);
    const date2 = new Date(timestamp2);

    if (isNaN(date1.getTime()) || isNaN(date2.getTime())) {
        throw new Error('Invalid timestamps');
    }

    if (date1 < date2) {
        return -1;
    } else if (date1 > date2) {
        return 1;
    } else {
        return 0;
    }
}

export const getConflictingEventsMap = (events: IEvent[]) => {
    const compareFunc = (e1: IEvent, e2: IEvent) => {
        return compareDates(e1.start_time, e2.start_time);
    }

    events.sort(compareFunc);

    const conflictingEventsMap: {
        [key: number]: number[]
    } = {}

    for(let i=0; i<events.length; i++) {
        for(let j=i+1; j<events.length; j++) {
            if (compareDates(events[i].end_time, events[j].start_time) === 1) {
                conflictingEventsMap[events[i].id] = conflictingEventsMap[events[i].id] ? [...conflictingEventsMap[events[i].id], events[j].id] : [events[j].id];
                conflictingEventsMap[events[j].id] = conflictingEventsMap[events[j].id] ? [...conflictingEventsMap[events[j].id], events[i].id] : [events[i].id];
            } else {
                break;
            }
        }
    }

    return conflictingEventsMap;
}
