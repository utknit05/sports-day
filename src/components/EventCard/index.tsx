import React from 'react';
import './styles.css';
import { TEventCardProps } from './types';
import { categoryGradientStyle, getButtonStyleClass } from './utils';
import { getDuration } from '../../utils/common';

const EventCard: React.FC<TEventCardProps> = (props) => {
    const handleOnClick = () => {
        if (props.type === 'selected') {
            props.onRemove(props.event);
            return;
        }

        if (props.event.disabled) {
            return;
        }

        props.onSelect(props.event);
    }

    return (
        <div className="eventCardContainer" style={categoryGradientStyle(props)}>
            <div className="eventDetailsContainer">
                <h1>{props.event.event_category[0].toUpperCase()}</h1>
                <div className="eventDescription">
                    <p className="eventName">{props.event.event_name}</p>
                    <p className="eventSubText">({props.event.event_category})</p>
                    <p className="eventSubText">{getDuration(props.event.start_time, props.event.end_time)}</p>
                </div>
            </div>
            <div className="eventButtonWrapper">
                <button
                    className={`eventActionButton ${getButtonStyleClass(props)}`}
                    onClick={handleOnClick}
                >
                    {props.type === 'selected' ? 'Remove' : 'Select'}
                </button>
            </div>
        </div>
    )
}

export default EventCard;
