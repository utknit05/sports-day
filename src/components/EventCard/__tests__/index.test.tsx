import EventCard from '..';
import * as cardUtils from '../utils';
import * as commonUtils from '../../../utils/common';
import { TEventCardProps } from '../types';
import { fireEvent, render, screen } from '@testing-library/react';

describe('EventCard Component', () => {
    const props: TEventCardProps = {
        event: {
            id: 3,
            event_name: "Freestyle 400M",
            event_category: "Swimming",
            start_time: "2022-12-17 15:00:00",
            end_time: "2022-12-17 16:00:00",
        },
        type: 'unselected',
        onSelect: jest.fn(),
    };

    beforeEach(() => {
        jest.spyOn(cardUtils, 'categoryGradientStyle').mockReturnValue({ backgroundImage: 'linear-gradient(to left, rgb(27, 214, 255), rgb(59, 157, 255))' });
        jest.spyOn(cardUtils, 'getButtonStyleClass').mockReturnValue('selectButton');
        jest.spyOn(commonUtils, 'getDuration').mockReturnValue('1 PM - 3 PM')
    })

    it('renders Event Card', () => {
        const { container } = render(<EventCard {...props} />);

        expect(container).toMatchSnapshot();
        expect(props.onSelect).toHaveBeenCalledTimes(0);

        const selectButton = screen.getByText('Select');

        fireEvent.click(selectButton);
        expect(props.onSelect).toHaveBeenCalledTimes(1);
    })
})
