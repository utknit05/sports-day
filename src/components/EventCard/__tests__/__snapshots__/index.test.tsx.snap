// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard Component renders Event Card 1`] = `
<div>
  <div
    class="eventCardContainer"
  >
    <div
      class="eventDetailsContainer"
    >
      <h1>
        S
      </h1>
      <div
        class="eventDescription"
      >
        <p
          class="eventName"
        >
          Freestyle 400M
        </p>
        <p
          class="eventSubText"
        >
          (
          Swimming
          )
        </p>
        <p
          class="eventSubText"
        >
          1 PM - 3 PM
        </p>
      </div>
    </div>
    <div
      class="eventButtonWrapper"
    >
      <button
        class="eventActionButton selectButton"
      >
        Select
      </button>
    </div>
  </div>
</div>
`;
